### PR TITLE
[s] Caps do_afters

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

Removes spamableness from the mech drill and caps the overall to 5 for futureproofing

## Why It's Good For The Game

Lag bad

## Changelog
:cl:
fix: Progress bars can no longer be spammed
/:cl:

### CODE REVIEW

File: code/modules/vehicles/mecha/equipment/tools/mining_tools.dm [hunk 1]
```diff
@@ -44,6 +44,9 @@
 					"<span class='userdanger'>[chassis] starts to drill [target]...</span>", \
 					 "<span class='hear'>You hear drilling.</span>")
 
+	if(source.do_afters)
```
Comment (Qustinnus) [inline code/modules/vehicles/mecha/equipment/tools/mining_tools.dm:47]: did you test this

Comment (TiviPlus) [reply code/modules/vehicles/mecha/equipment/tools/mining_tools.dm:47]: yes

**Comment by Qustinnus** [comment]
instead of capping the amount of do_afters(), do_afters() should instead use a source system most likely, so you cant have multiple do_afters from the same source, by making do_afters into an assoc list.

File: code/__HELPERS/mobs.dm [hunk 2]
```diff
@@ -250,6 +250,9 @@ GLOBAL_LIST_EMPTY(species_list)
 	if(target && !isturf(target))
 		Tloc = target.loc
 
+	if(LAZYLEN(user.do_afters) >= 5)
```
Comment (Jared-Fogle) [inline code/__HELPERS/mobs.dm:253]: Define.

File: code/modules/vehicles/mecha/equipment/tools/mining_tools.dm (hunk 1 above)
Comment (Tlaltecuhtli) [reply code/modules/vehicles/mecha/equipment/tools/mining_tools.dm]: does it work

**Comment by Ryll-Ryll** [comment]
> instead of capping the amount of do_afters(), do_afters() should instead use a source system most likely, so you cant have multiple do_afters from the same source, by making do_afters into an assoc list.

That is how it works, see the macro IS_INTERACTING_WITH

**Comment by Rohesie** [comment]
If this happens, it'd be nice if it was integrated with #54409 , so you can set flags to ignore the cap.

**Comment by ShibaInuLord** [comment]
no more breaking out of locker rainbow 0/10

File: code/__HELPERS/mobs.dm [hunk 3]
```diff
@@ -258,6 +258,9 @@ GLOBAL_LIST_EMPTY(species_list)
 	if(target && !isturf(target))
 		target_loc = target.loc
 
+	if(!(timed_action_flags & IGNORE_DO_AFTER_CAP) && LAZYLEN(user.do_afters) >= 5)
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm:261]:
```suggestion
	if(!(timed_action_flags & IGNORE_DO_AFTER_CAP) && LAZYLEN(user.do_afters) >= MAX_DO_AFTERS)
```

Comment (Rohesie) [reply code/__HELPERS/mobs.dm:261]: This check could go before the `target_loc` variable assignment, really.

**Review by Rohesie — CHANGES_REQUESTED** [review]
Could you add this behavior, for consistency, to all the do_acter-like procs? `/proc/do_atom`, `/proc/do_after_mob`, `/proc/do_mob` and `/proc/do_after`

**Comment by Cyberboss** [comment]
Spam is bad yes, but a global cap seems like a bad idea(tm)

**Comment by Rohesie** [comment]
I also believe this should be handle on a per-case basis.
Some do_afters are not supposed to be cancelled, and they might get caught by the cap.

**Comment by Qustinnus** [comment]
to bump what I said: It does not use an assoc list now and probably should, saving interactions from specific strings similar to traits, and having an int to keep track of the amount of current interactions going on.

**Comment by LemonInTheDark** [comment]
Poke?

**Pull request closed without merging**